Proposal - Pod safety and termination guarantees

smarterclayton opened the pull request:
This proposal describes how to evolve the Kubernetes cluster to provide
at-most-one semantics for pod execution and to allow selective
relaxation where necessary to heal partitions.

This is a continuation of pod graceful deletion and completes the safety guarantees begun in that change (https://github.com/kubernetes/kubernetes/issues/1535)

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34160)

<!-- Reviewable:end -->

smarterclayton committed:
Proposal - Pod safety and termination guarantees

This proposal describes how to evolve the Kubernetes cluster to provide
at-most-one semantics for pod execution and to allow selective
relaxation where necessary to heal partitions.

smarterclayton commented:
@bprashanth @saad-ali @thockin @vishh

smarterclayton commented:
Linked a number of issues that relate to pet safety, pod termination guarantees, and storage safety.

gmarek commented:
cc @wojtek-t @gmarek

AdoHe commented:
/cc @AdoHe

fabiand commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+entity will be spawned) but may become unavailable (cluster no longer has
+a sufficient number of members). The Pet Set guarante must be strong enough
+for an administrator to reason about the state of the cluster by observing
+the Kubrenetes API.
```
Typo: `Kubernetes`

caesarxuchao commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+
+Clients today may assume that force deletions are safe. We must appropriately
+audit clients to identify this behavior and improve the messages. For instance,
+`kubectl delete --grace-period=0` could print a warning and require `--confirm`:
```
If the pod has pending finalizers, --grace-period=0 will not delete the pod from the key-value store. Rather than overloadin the `grace-period` flag, can we define a new flag, like `--forceful-deletion`, which will instruct the API server to also ignore pending finalizers?

smarterclayton replied:
I don't think of force deletion as bypassing finalizers (in the context the proposal is written).  This would be delete 0 as it is today.  If that bypasses finalizers then we may want to clarify what exactly we want the role of finalizers to be wrt deletion here so it's clear.

smarterclayton replied:
I don't think of force deletion as bypassing finalizers (in the context the proposal is written).  This would be delete 0 as it is today.  If that bypasses finalizers then we may want to clarify what exactly we want the role of finalizers to be wrt deletion here so it's clear.

smarterclayton commented:
Gentle nag for more review from those this impacts - this is a p1 1.5 item for petsets

smarterclayton commented:
@erictune

marun commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+5. The kubelet on node `A` observes the pod references a PVC that specifies RWO which
+   requires "attach" to be successful
+6. The attach/detach controller observes that a pod has been bound with a PVC that
+   requires "attach", and attempts to execute a CAS update on the PVC/PV attaching
```
Curious, what is 'CAS'?

bprashanth replied:
https://en.wikipedia.org/wiki/Compare-and-swap

marun reviewed (commented):
spelling/grammar nits

bprashanth commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+  to communicate those implications.
+* All existing controllers in the system must be limited signaling pod
+  termination (starting graceful deletion), and are not allowed to force
+  delete a pod.
```
why not just leave this upto forgiveness? pods get default forgiveness that does what we do today (tolerate partition for 5m). All controllers are modified to respect forgiveness, users can modify it, petset controller *will* modify it to favor consistency by default.

smarterclayton replied:
Force deletion is what causes split brain.  The only guarantee that can ensure the process has been terminated (and thus is safe to launch another) is the kubelet or a fencer.  So only the kubelet, a fencer, or a human should ever force delete.  The petset controller shouldn't have to apply any consistency guarantee.

smarterclayton replied:
Forgiveness is delay before delete.  Should have no coupling to force deletion.

davidopp replied:
Yeah, and forgiveness is a delay before a delete *that is triggered by placement of a taint* so there are plenty of deletion scenarios that forgiveness wouldn't help with. (It would help with node unreachable (Unknown node condition), though.)

bprashanth replied:
I think we're agreeing (I forgot about deletion timestamp). You're suggesting that we use a combination of forgiveness + disruptionBudget to communicate when to delete to the fencer right?

smarterclayton replied:
Yes, will add a paragraph in fencing for this.

bprashanth commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+
+The changes above allow Pet Sets to ensure at-most-one pod, but provide no
+recourse for the automatic resolution of cluster partitions during normal
+operation. For that, we propose a **fencing controller** which exists above
```
I think there's a gap between the previous section (which breaks current AP replica-set behavior by preventing nodecontroller delete) and this (which only fences). What deletes pods in this world?  are we saying it's *always* a human in the partition case?

smarterclayton replied:
A cloud provider can properly observe dead nodes and delete them, so we can preserve that aspect  Partition only prevents pods from being removed from etcd.  I think the default behavior has to be to never induce split brain except under stronger guarantees than "I waited".

It's possible that we could extend forgiveness to be an upper bound of tolerance for failure, and so could be opt in to induce a split brain for pet sets.  But that's not safe without either unique fenced storage or code inside the petset that can tolerate membership changes by removing the items with a certain clock window.

smarterclayton replied:
is there a forgiveness node variant we can create that allows both "preserve without deleting" and "preserve without force deleting"?

davidopp replied:
Forgiveness hasn't been fully implemented yet but I was assuming that once the forgiveness period expires, it would delete the same way node controller deletes today when there's a NotReady (bad node/kubelet) or Unknown (node unreachable), i.e. the code here
https://github.com/kubernetes/kubernetes/blob/4bda90e89c1e945c74856553b47a517a3244f874/pkg/controller/node/controller_utils.go#L45
which IIUC does a regular delete not a force delete? (Where is the client Delete() operation that it's using defined?)

bprashanth replied:
isn't there going to be a `event: *, duration: infiinte` flavor for forgiveness?

smarterclayton replied:
@davidopp https://github.com/kubernetes/kubernetes/blob/4bda90e89c1e945c74856553b47a517a3244f874/pkg/controller/node/controller_utils.go#L95 is invoked after timeout period.  The proposal calls for the forceful delete to be removed.

davidopp replied:
Once we start using taints/tolerations/forgiveness to control eviction due to unreachable/not ready node (unfortunately it was decided this will not be enabled by default in 1.5, though it will be implemented) you will be able to say "stay bound to the node no matter what happens" . I think forcefullyDeleteNode() which Clayton pointed to may still be OK since it only activates for a node that has disappeared from the cloud provider, and IIRC you said it's fine for nodes the cloud provider knows is gone, to be deleted? There may be other places in the NC where we force delete, and presumably we would have to replace those with graceful delete?

smarterclayton replied:
Yes

bprashanth commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+    update on the PVC/PV clearing its attach state.
+14. The attach/detach controller observes the second pod has been scheduled and
+    attaches it to node `B` and pod 2
+15. The kubelet on node `B` observes the attach and allows the pod to execute.
```
I think we already have this flow working on cloudprovider backed storage, we just need to:    
1. teach *all* storage plugins (even those that are not "attachable") to honor it
2. communicate between scheduler and volume controller about where exactly to attach (that's the first half of https://github.com/kubernetes/kubernetes/pull/30044)

The existing flow uses 2 fields in node.Status:
  * volumesInUse: To block volumecontroller from detaching a mounted volume
  * volumesAttached: To block another node from attaching a pre-attached volume

I can think of 2 things we might need to fix off the top of my head:    

1. I think there's a gap with the storage plugin on the node, where some allow 2 pods to bind mount the same volume RWO. I think gce allows this currently.
2. Today, some volume controllers time out and force detach after a "long" time (2h). I think this needs fixing, especially if we're going to use them as described.

@kubernetes/sig-storage

smarterclayton replied:
Node status isn't transactional to the pv, so it doesn't prevent races when claiming.

bprashanth replied:
I think the scheduler should always make the assignment, not volume controller. Given that scheduler needs to write hostname *somewhere*, the volumecontroller just needs to trigger its normal attach flow on that. Maybe you meant active/passive volume controller setup ?

smarterclayton replied:
I would argue that all the attach logic today observes scheduler decisions, and that except for maybe the storage allocation for local persistent volume, it's more natural for attach detach to continue to observe scheduling and guarantee locks.  It seems to match the description of what attach_detach_controller owns (binding PVs to nodes transactionally by making API calls to either cloud or kube)

thockin replied:
The fact that a PV is attached to a node isn't clearly denoted in a way that is helpful to this idea, I think.  There's some real work needed to make sure this is safe.  Given how tricky the existing binding and attach/detach code is...

bprashanth replied:
Yeah thats the same problem we need to solve for sticky emtpy dir. We discussed using a field or labels on that proposal.

bprashanth commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+
+## Backwards compatibility
+
+On an upgrade, pet sets would not be "safe" until the above behavior is implemented.
```
We'd just use (i.e document) finalizers as poor mans forgiveness, right?

smarterclayton replied:
Ah.  Do you think we'd need it?

smarterclayton commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+  grace period and performs the following actions:
+  * Executes the pre-stop hook, if specified, waiting up **grace period**
+    before continuing
+  * Sends the termination signal to the container runtime (SIGTERM)
```
Yes, we invoke docker stop which calls stopsignal on the image.

foxish commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+  termination (starting graceful deletion), and are not allowed to force
+  delete a pod.
+  * The node controller will no longer be allowed to force delete pods -
+    it may only signal deletion
```
Aren't force deletions from the node controller still needed for the guarantee that RS/RC provide, that at least N replicas are running?

smarterclayton replied:
RS/RC *immediately* spin up new pods as soon as the node controller signals termination on the pod (by starting graceful delete).  They exclude any terminating pods from their count.

@kargakis raised today that that behavior actually violates the implicit guarantee of the recreate deployment strategy - that there is a period where there are zero pods running of the old version.  I think we need to resolve that independently and he was filing an issue.

If we treat node deletion as implicit unlock (it probably is) then both pod GC and node controller would be allowed to force delete pods.  However, I'm not 100% sure since today there are cases where the node would just recreate its node object, and people may not be expecting that delete node means "hard node evacuate".

bprashanth commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,388 @@
+    terminated.
+* Application owners must be free to force delete pods, but they *must*
+  understand the implications of doing so, and all client UI must be able
+  to communicate those implications.
```
these 2 are already true today right? (assumign "normal" operation is not netsplit)

smarterclayton replied:
Yes

smarterclayton replied:
I think the "implications" are not explained.

bprashanth commented:
Is the 1.5 MVP to delete code that force-deletes pods in node and other controllers (as described: https://github.com/kubernetes/kubernetes/issues/31762#issuecomment-250595174)? Out of the 3 sections, the first is the one that solves the core issue. Both fencer and RWO lock border new features.

smarterclayton commented:
Yes, and to add --force for kubectl delete (0 will not be allowed by
default or made to be 1 under the covers).

Fencer and RWO can be solved in a future release, we would document how to
be safe and what admins have to do to be safe.

I'll sign up for all of the work here, I have time set aside for it anyway
(or I can review if someone is already assigned).

On Oct 12, 2016, at 8:36 PM, Prashanth B <notifications@github.com> wrote:

Is the 1.5 MVP to delete code that force-deletes pods in node and other
controllers (as described: #31762 (comment)
<https://github.com/kubernetes/kubernetes/issues/31762#issuecomment-250595174>)?
Out of the 3 sections, the first is the one that solves the core issue.
Both fencer and RWO lock border new features.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-253378824>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p4yODD1qRvlwsz_5dw6jzs4M0-95ks5qzX0QgaJpZM4KPd1x>
.

bprashanth commented:
SGTM, implementation should be straightforward. More importantly, doing this will enable us to test some basic split brain. Running by @erictune (do you still think we need a "Lost" state?)

smarterclayton committed:
Review 1: Comments and new question

smarterclayton commented:
Added some more content based on the discussion AND added one more question - should node deletion be a "client can assume pod processes are stopped".  Answer is no (I think), which means pod GC controller cannot force delete pods.

If it could, then the following scenario would lead to pet set split brain:

1. User has pod 1 on node 1
2. Node 1 is partitioned for 30 minutes
3. During partition, admin deletes node 1
4. During partition, PodGC observes missing node and triggers graceful then forceful deletion
5. New pet is provisioned on node 2
6. Two copies of pod are running on cluster at the same time.

The implication there is there is no automated way to force delete pods that are partitioned without a fencer, which means admins are responsible for .  We could potentially add a field selector for get that allows you to select partitioned pods, and then support `kubectl delete pods --all-namespaces --all --force --field=status.conditions[*].partitioned` or something.

smarterclayton commented:
Since @vishh is on vacation @dchen1107 @thockin @erictune @bgrant0607 it would be good to have a node / storage / cluster consistency opinion on this.  TL-DR: only "safe" clients who can verify process are stopped should force delete pods - safe is kubelet, fencer, and human.  Everyone else requests deletion.

bprashanth commented:
I was operating on the assumption that if the admin deletes a node (either from cloud or kubernetes) it means they have taken the necessary steps to properly fence it. We would add a warning that said: 

PetSet identity guarantees can be broken if:        
1. You have a custom controller that force deletes
2. You delete a node that is still running a pet without properly shutting it down/fencing it

To fix 1, we could put in something like a generic pod/kubelet finalizer mechanism post 1.5. 
To fix 2, we would need a "kubectl fence node" or something to that effect, but since we won't get that in 1.5, should we just document?

smarterclayton commented:
Yeah we can document.  I think 2 is the comment I made about a way for an
admin to request fencing, which is probably setting a bit on the node that
indicates that the node is lost / down / unavailable that the fencer can
use as an unambiguous signal.  The fencer can then try to fence the node -
if it succeeds, it can take action and force delete pods or force delete
the node.  If it fails, it reports a condition (maybe).

On Fri, Oct 14, 2016 at 1:32 PM, Prashanth B <notifications@github.com>
wrote:

> I was operating on the assumption that if the admin deletes a node (either
> from cloud or kubernetes) it means they have taken the necessary steps to
> properly fence it. We would add a warning that said:
>
> PetSet identity guarantees can be broken if:
>
> 1. You have a custom controller that force deletes
> 2. You delete a node that is still running a pet without properly shutting
> it down/fencing it
>
> To fix 1, we could put in something like a generic pod/kubelet finalizer
> mechanism post 1.5.
> To fix 2, we would need a "kubectl fence node" or something to that
> effect, but since we won't get that in 1.5, should we just document?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-253868289>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p8W8kkhb-w_OXKrF-ooGW5P5KFIaks5qz7y4gaJpZM4KPd1x>
> .
>

foxish commented:
From my looking over the node controller, I found the following cases in which we are force-deleting pods.

1. Pods on nodes which have not responded in a long time. (the case we covered, in case of network partitions etc)
2. Terminating pods which are *not scheduled* yet.
3. Pods which have been scheduled on a non-existent node.
4. Terminating pods on nodes running kubelet version < 1.1.0 which do not support graceful termination.

Case #2 and #3, it appears that the kubelet cannot handle the deletion, because there is no kubelet that is responsible at that point.
#4 seems like a special case.

@smarterclayton @bprashanth @mikedanese what do you think?

cc @erictune

bprashanth commented:
Yeah that's what i expected:     

1. Normal case
2. You can't easily erase spec.nodeName, so assuming it was never actually on any node, and it's terminating now, even if spec.nodeName is suddenly set the kubelet shouldn't pick it up at the wrong time. So I think deleting in this case is safe.
3. We already established that if the node doesn't show up in kubectl or cloudprovider we consider it "fenced" (the race is: if you are placed on some node, the node doesn't exist in `kubectl get no` or `cloud-provider get no`, the nodecontroller decides to delete the pod, then suddenly the node shows up, the kubelet grabs the pod, starts running it, and gets partitioned from the nodecontroller at exactly the wrong moment. I don't think this should happen unless the user sets spec.nodeName *before* creating the node).
4. In theory we only really support 2 versions back. IMO it would be fine to just document the caveat (fencing won't work in node versions < 1.1).

smarterclayton commented:
PodGC should handle 1 and 3 (that's the last question).  We can't tell the
difference between 3 when an admin deletes the node or when a user directly
passes nodeName, so we have to assume the former.

2 is acceptable only when we have preconditions, which we do (resource
version on bind).  We may have implemented graceful deletion so that
unscheduled pods are immediately deleted anyway (which we can do because
nodeName is one way set).  Will double check.

4 at this point will be to tell the admin to upgrade (as it applies to this
proposal).

On Oct 17, 2016, at 9:49 PM, Anirudh Ramanathan <notifications@github.com>
wrote:

From my looking over the node controller, I found the following cases in
which we are force-deleting pods.

   1. Pods on nodes which have not responded in a long time. (the case we
   covered, in case of network partitions etc)
   2. Terminating pods which are *not scheduled* yet.
   3. Pods which have been scheduled on a non-existent node.
   4. Terminating pods on nodes running kubelet version < 1.1.0 which do
   not support graceful termination.

Case #2 <https://github.com/kubernetes/kubernetes/issues/2> and #3
<https://github.com/kubernetes/kubernetes/issues/3>, it appears that the
kubelet cannot handle the deletion, because there is no kubelet that is
responsible at that point.
#4 <https://github.com/kubernetes/kubernetes/issues/4> seems like a special
case.

@smarterclayton <https://github.com/smarterclayton> @bprashanth
<https://github.com/bprashanth> @mikedanese <https://github.com/mikedanese>
what do you think?

cc @erictune <https://github.com/erictune>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-254384363>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p3P-Vu42zqO2HPrB-L6vd8N8vdXfks5q1CW8gaJpZM4KPd1x>
.

foxish commented:
Added another e2e test case to https://github.com/kubernetes/kubernetes/pull/34045, to ensure that the pet is not deleted during a network partition. 
Now we can make the changes to the node controller; I think this will also need modifying existing e2e tests which expect the node controller eviction behavior. 4 has been dealt with, so, in summary, the actions required for the other items would be:

1. The node controller change to not force-delete would make the above test pass.
2. The unscheduled pods which are terminating [get deleted immediately](https://github.com/kubernetes/kubernetes/blob/fd874ab1731b082115c8cb206d41b15221de1fb4/pkg/registry/core/pod/strategy.go#L115). 
3. The PodGC controller [already performs](https://github.com/kubernetes/kubernetes/blob/fd874ab1731b082115c8cb206d41b15221de1fb4/pkg/controller/podgc/gc_controller.go#L173) force-deletions of "orphaned pods" (bound to non-existent nodes), which means that same logic is redundant in the node controller, and can be removed.
4. Will be documented.

At the end of the above, we should have just PodGCController doing force deletions.

davidopp commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,419 @@
+    container image's STOPSIGNAL on Docker)
+  * Waits 2 seconds, or the remaining grace period, whichever is longer
+  * Sends the force termination signal to the container runtime (SIGKILL)
+* Once the kubelet observes the container is fully terminated, it issues
```
Note that this contradicts the comment on DeletionTimestamp in the API, which says "The Kubelet will react by sending a graceful termination signal to the containers in the pod. Once the resource is deleted in the API, the Kubelet will send a hard termination signal to the container."

Please fix whichever is wrong. (Unless the steps you wrote above are a proposal, not what is currently implemented, in which case please make that clear.)

smarterclayton replied:
The API comment is inaccurate.  The description in the proposal is the currently implemented behavior.

smarterclayton replied:
Fixed in https://github.com/kubernetes/kubernetes/pull/35481

vishh replied:
nit: "deleted" here refers to containers reaching a terminal state. The pod is still not technically "deleted"

davidopp commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,419 @@
+ReplicaSets and ReplicationControllers both attempt to preserve availability
+of their constituent pods over ensuring **at most one** semantics. So a
+replica set to scale 1 will immediately create a new pod when it observes an
+old pod has been deleted, and as a result at many points in the lifetime of
```
maybe say "requested to be deleted" to clarify that you mean "deletion process initiated" rather than "gone"?

smarterclayton replied:
Yeah, we are unclear on terminology.  Technically graceful is deletion, but it would be more clear to separate action - beginning graceful deletion - from result - API resource deleted.

davidopp commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,419 @@
+* Deleting a node from etcd can be seen as a signal to the cluster that
+  the node is permanently partitioned. We must audit existing components
+  to verify this is the case.
+  * Alternatively, we could require that pods must be individually
```
Saying you can only delete a node when all pods on it are gone might be reasonable (though we have several flavors of pods that can't really be deleted, unfortunately).

gmarek replied:
Currently nothing deletes Daemons, and it's impossible to delete mirror Pods AFAIK.

smarterclayton commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,419 @@
+### Guarantees provided by replica sets and replication controllers
+
+ReplicaSets and ReplicationControllers both attempt to preserve availability
+of their constituent pods over ensuring **at most one** semantics. So a
```
Sure.

smarterclayton commented:
> Once we start using taints/tolerations/forgiveness to control eviction due to unreachable/not ready node (unfortunately it was decided this will not be enabled by default in 1.5, though it will be implemented) you will be able to say "stay bound to the node no matter what happens" .

So infinite timespan of forgiveness?

>  I think forcefullyDeleteNode() which Clayton pointed to may still be OK since it only activates for a node that has disappeared from the cloud provider, and IIRC you said it's fine for nodes the cloud provider knows is gone, to be deleted? 

Since we now have pod GC controller we don't need the node controller to provide this guarantee anymore.  Deleting the node in the API (absence of the node) could be an adequate signal for permanent partition.  So if the pod GC controller observes a pod with no corresponding node then force deletion is acceptable, in which case node controller only has to delete the node when cloud provider signals.  This may be subject to race conditions though without some ordering between nodes and pods - just looking at the pod gc controller logic today, if the node watch lagged it's possible that the pod gc controller would delete pods on a newly scheduled node.

This I think is the last outstanding question here.

> There may be other places in the NC where we force delete, and presumably we would have to replace those with graceful delete?

All of the places in NC should have started with graceful then force after observed time period (since 1.1).

gmarek commented:
We had podGC for few releases now.

In all places in NC, except force deletion Pods from non-existing Nodes, we're starting with graceful deletion.

gmarek commented:
I'm worried that this proposal removes some of the automation we have in place now, and requires human action for pretty basic stuff - in large clusters (espacially on-prem), Nodes will die, and it won't be a rare event.

foxish commented:
@gmarek I think we agree that in case of cloud providers (where we can make an API call to figure out if a node has gone down), the proposed alteration in behavior is safe, because if we did see that a node has gone down permanently, the NC deletes the node, and the podGC force-deletes the pods thereafter ([podgc/gc_controller.go:172](https://github.com/kubernetes/kubernetes/blob/6079053407705212c8cc3f039e8b767c6bccd0a0/pkg/controller/podgc/gc_controller.go#L172)).

Your concern as I understand it, is that on-prem, there is no direct API to figure out whether a machine has gone down. I would argue that the "automation" that exists, in cleaning up pods from unreachable nodes has the potential to be incorrect and should be made explicit anyway. If as we do currently, we do not differentiate between:

1. kubelet being unreachable on a node that is still running the user's pods.
2. machine going down for maintenance and no longer running the user's pods.

As of now, our action after 300 missed heartbeats is the exact same in either case; we force delete the pods from the api-server. However, if it was case 1, this does not reflect reality. The pods should not be deemed deleted unless we are certain that they are not running on that node which has for some reason, stopped communicating. In my opinion, we should instead make the recommendation that on-prem users of large clusters write a reconciler which keeps the machines that are up, in sync with the node resources in kubernetes. It would be their responsibility to ensure that a node has really gone down, and clean it up from kubernetes, and until then, it would simply be in the `NodeNotReady` state, which is as much information as we have about it. In summary, I think it should be the on-prem cluster administrator's responsibility to ensure that the kubernetes node resources match up with their underlying infrastructure.

kfox1111 commented:
for bare metal, ipmi is an option for a lot of sites. If the node is suspected down, and a critical load was running on it, it can be ipmipowered off to fence the node.

foxish commented:
@gmarek, For 1.5, we are pretty much doing the first step of https://github.com/kubernetes/kubernetes/issues/20497#issuecomment-190976488.

davidopp commented:
This change seems fine with me. Among currently-existing controllers, the new behavior will only apply to PetSet, because the others create a replacement pod as soon as deletionTimestamp gets set. Given that this safety requirement is part of the definition of PetSet, I think treating it specially is reasonable. Also, this change will eliminate the current behavior where controller-less pods disappear with no trace when their node goes down for more than 5 minutes, and that seems like an improvement.

(The above were pointed out by @erictune and @bprashanth)

foxish commented:
The action as per this proposal is to move some of the functionality provided by the NC, i.e. force-deleting pods when a node resource is deleted from kubernetes, into the podGC. However, from the code, I see that:
The podGC controller may not always be running since whether it runs at all is [governed by TerminatedPodGCThreshold](https://github.com/kubernetes/kubernetes/blob/b471398f1f0d5908fee6dbdac06ebbeeefa34faf/cmd/kube-controller-manager/app/controllermanager.go#L274), which is a flag passed to the controller manager. 

> `--terminated-pod-gc-threshold value` Number of terminated pods that can exist before the terminated pod garbage collector starts deleting terminated pods. If <= 0, the terminated pod garbage collector is disabled. (default 12500). 

If we were to move the pod deletion in case of node resource deletion into the podGC, users who run kubernetes with the podGC disabled for some reason may see behavioral changes. By the looks of it, the podGC has evolved to do much more than just cleaning up terminated pods beyond a certain threshold number. Example: https://github.com/kubernetes/kubernetes/pull/32495. I think we should always be running the PodGC controller and only gate the logic that cleans up terminated pods behind `TerminatedPodGCThreshold`. @gmarek Does that seem reasonable?

smarterclayton commented:
We added node clean up pretty early on, so it is appropriate that it be always on.  Will note that in the proposal.

smarterclayton committed:
Review 2: Additional comments

smarterclayton commented:
Updated with all comments to date.

smarterclayton commented:
CLI changes in #35484

fabiand commented:
@kfox1111 there are a couple of fencing providers for different solutions. It could actually be a separate proposal to wrap those in a separate bare-metal provider.

smarterclayton commented:
Yup, Mike was right for PetSets.

On Oct 24, 2016, at 6:25 PM, Anirudh Ramanathan <notifications@github.com>
wrote:

@gmarek <https://github.com/gmarek>, For 1.5, we are pretty much doing the
first step of #20497 (comment)
<https://github.com/kubernetes/kubernetes/issues/20497#issuecomment-190976488>
.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-255883743>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p6fNGW5HBUHZgU0u0IoC2KBAwsKPks5q3TA8gaJpZM4KPd1x>
.

smarterclayton commented:
Plus the design of a fencer abstraction.

On Oct 24, 2016, at 6:25 PM, Anirudh Ramanathan <notifications@github.com>
wrote:

@gmarek <https://github.com/gmarek>, For 1.5, we are pretty much doing the
first step of #20497 (comment)
<https://github.com/kubernetes/kubernetes/issues/20497#issuecomment-190976488>
.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-255883743>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABG_p6fNGW5HBUHZgU0u0IoC2KBAwsKPks5q3TA8gaJpZM4KPd1x>
.

smarterclayton commented:
Check over the list in the proposal of tools for fencing - if there is
anything I missed I need to add it.

On Mon, Oct 24, 2016 at 4:05 PM, kfox1111 <notifications@github.com> wrote:

> for bare metal, ipmi is an option for a lot of sites. If the node is
> suspected down, and a critical load was running on it, it can be
> ipmipowered off to fence the node.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-255850162>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p-4L_ekk70mIT-cGLyYCNwTdQvzrks5q3Q-JgaJpZM4KPd1x>
> .
>

smarterclayton commented:
Yes.  In the proposal "fencer" is a placeholder for "reconciler between
reality and cluster stable state".  For an on-prem cluster, most large
organizations already have tooling that can perform this reconciliation,
and administrators are paged / notified to make final decisions (or
automated decisions).  In messaging this in 1.5, we would make it clear
that the process for handling dead nodes on-prem is:

1. Identify unambiguously that the node is dead OR Take manual action to
ensure it is dead
2. Delete the node or force delete the pods on the node (depending on the
action taken above, and the provisioning strategy).

In future releases we can explore one or more fencing implementations that
perform these actions (or describe how to implement them using existing on
prem tools).


On Mon, Oct 24, 2016 at 3:23 PM, Anirudh Ramanathan <
notifications@github.com> wrote:

> @gmarek <https://github.com/gmarek> I think we agree that in case of
> cloud providers (where we can make an API call to figure out if a node has
> gone down), the proposed alteration in behavior is safe, because if we did
> see that a node has gone down permanently, the NC deletes the node, and the
> podGC force-deletes the pods thereafter (podgc/gc_controller.go:172
> <https://github.com/kubernetes/kubernetes/blob/6079053407705212c8cc3f039e8b767c6bccd0a0/pkg/controller/podgc/gc_controller.go#L172>
> ).
>
> Your concern as I understand it, is that on-prem, there is no direct API
> to figure out whether a machine has gone down. I would argue that the
> "automation" that exists, in cleaning up pods from unreachable nodes has
> the potential to be incorrect and should be made explicit anyway. If as we
> do currently, we do not differentiate between:
>
>    1. kubelet being unreachable on a node that is still running the
>    user's pods.
>    2. machine going down for maintenance and no longer running the user's
>    pods.
>
> As of now, our action after 300 missed heartbeats is the exact same in
> either case; we force delete the pods from the api-server. However, if it
> was case 1, this does not reflect reality. The pods should not be deemed
> deleted unless we are certain that they are not running on that node which
> has for some reason, stopped communicating. In my opinion, we should
> instead make the recommendation that on-prem users of large clusters write
> a reconciler which keeps the machines that are up, in sync with the node
> resources in kubernetes. It would be their responsibility to ensure that a
> node has really gone down, and clean it up from kubernetes, and until then,
> it would simply be in the NodeNotReady state, which is as much
> information as we have about it. In summary, I think it should be the
> on-prem cluster administrator's responsibility to ensure that the
> kubernetes node resources match up with their underlying infrastructure.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-255839416>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p2sfByDZ4-3FCS9V3RqwN0Y5d6gmks5q3QXHgaJpZM4KPd1x>
> .
>

smarterclayton commented:
Cloud Provider should delete dead nodes, correct?  Large on-prem clusters
are almost certainly more broken with force deletion (in terms of impact to
operations) than non force deletion.

On Mon, Oct 24, 2016 at 4:34 AM, Marek Grabowski <notifications@github.com>
wrote:

> I'm worried that this proposal removes some of the automation we have in
> place now, and requires human action for pretty basic stuff - in large
> clusters (espacially on-prem), Nodes will die, and it won't be a rare event.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-255679890>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p6IosxrjocgaBJoGPnTKi9c_4ZqXks5q3G2ZgaJpZM4KPd1x>
> .
>

gmarek commented:
@smarterclayton - I spoke with @foxish earlier today. The main take-away is that this would be a change to the "old" behavior, so it's not fully backward compatible, thus it's a more product decision, than technical one and I'll let people that talk to users decide.

smarterclayton commented:
So your technical concern is that by making this change without a fencer,
we don't keep the number of terminated pods in the system low during node
partitions?

On Tue, Oct 25, 2016 at 2:48 PM, Marek Grabowski <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> - I spoke with @foxish
> <https://github.com/foxish> earlier today. The main take-away is that
> this would be a change to the "old" behavior, so it's not fully backward
> compatible, thus it's a more product decision, than technical one and I'll
> let people that talk to users decide.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/34160#issuecomment-256136739>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p8CM6a5-xj7S6h3oJfHurTpE4Osqks5q3k7_gaJpZM4KPd1x>
> .
>

gmarek commented:
I'm not sure it can be called technical concern, but yes (with the slight correction - not "terminated", but "terminating")

thockin commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
```
pending PR to remove all this gunk

thockin commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+By leveraging existing concepts, we can allow the scheduler and the kubelet to enforce
+a guarantee that an RWO volume can be used on at-most-one node at a time.
+
+In order to properly support region and zone specific storage, Kubernetes adds node
```
@jsafrane @pmorie @saad-ali @kubernetes/sig-storage

smarterclayton commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+  * Sends the force termination signal to the container runtime (SIGKILL)
+* Once the kubelet observes the container is fully terminated, it issues
+  a status update to the REST API for the pod indicating termination, then
+  issues a DELETE with grace period = 0.
```
The Kubelet doesn't need a finalizer, because the finalizer doesn't do anything.  Force delete has to trump finalizers (so app admins and fencers can resolve partitions), and since both fencers and app admins can trump finalizers, the finalizer is only symbolic and adds no value.

smarterclayton commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+
+* Give the Kubelet sole responsibility for normal deletion of pods -
+  only the Kubelet in the course of normal operation should ever remove a
+  pod from etcd (only the Kubelet should force delete)
```
Sure

smarterclayton commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+another node and have it join the cluster safely (pets 0 and 1 constitute a
+quorum for membership change).
+
+This proposal does not alter the behavior of finalizers - instead, it makes
```
Comments above, maybe we're operating under different definitions of finalizers.

smarterclayton commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+and to take corresponding steps to prevent a dead machine from coming back
+to life automatically.
+
+Fencing controllers most benefit services that are not innately replicated
```
There can be controllers that fence pods, fence nodes, fence nodes and pods, fence networks, fence storage, etc.  I can list that those are a possibility.

smarterclayton commented:
> The fact that a PV is attached to a node isn't clearly denoted in a way that is helpful to this idea, I think. There's some real work needed to make sure this is safe. Given how tricky the existing binding and attach/detach code is...

Agree - but I think that's something we have to solve, which is why it's here.  A fencer must have access to some of that info, if the storage is managed by the cluster.  And attach / detach has to be atomic to something in order for a fence to even be possible.

apsinha commented:
@gmarek 

> @smarterclayton - I spoke with @foxish earlier today. The main take-away is that this would be a change to the "old" behavior, so it's not fully backward compatible, thus it's a more product decision, than technical one and I'll let people that talk to users decide.

It looks like this change would be valuable to the majority of our users and improves the general correctness of the system. We should document the change in behavior thoroughly, with special mention for bare metal on-premise users wanting to upgrade to 1.5 and beyond.

gmarek commented:
Thanks @apsinha - SGTM.

smarterclayton commented:
Outstanding issue now is finalizer vs no - whether it adds anything (so we can wait to add it for a real need), whether there's an edge case we haven't considered (or whether is obscured by other decisions).

smarterclayton commented:
@thockin we can probably merge this and come back to the finalizer question in a follow up.  For 1.5 everything we need is captured (and merged).

smarterclayton commented:
If there are no further comments I'd like to merge this and discuss finalizers in a follow up to this issue.  No one has articulated a blocker on this yet.  This is still advisory for storage - i think we'll have to expand this with a storage related topic.

smarterclayton commented:
Hrm, I need to fix the pet set references and a few wording changes.

derekwaynecarr commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+* Give the Kubelet sole responsibility for normal deletion of pods -
+  only the Kubelet in the course of normal operation should ever remove a
+  pod from etcd (only the Kubelet should force delete)
+  * The kubelet must not delete the pod until all processes are confirmed
```
do you want volume clean-up to happen prior to deletion from the apiserver as well?

smarterclayton replied:
It depends - certainly having attach/detach controller able to observe that on pod deletion helps (once the pod is gone, detach controller has to start trying to detach anyway, so might as well make that consistent).

vishh replied:
I'd prefer cleaning up of all resources belonging to a pod before the corresponding object gets deleted.

vishh replied:
@dashpole ^^

derekwaynecarr commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+  to communicate those implications.
+  * Force deleting a pod may cause data loss (two instances of the same
+    pod process may be running at the same time)
+* All existing controllers in the system must be limited to signaling pod
```
I am wondering if we can protect this in code somehow.  For example, separate `ForceDelete` from `Delete`, so `Delete` does not let you do a force delete so we can minimize reviewer burden in the future.

smarterclayton replied:
It's not terribly difficult to audit today in our code base, but I agree that some level of protection / review is important.

derekwaynecarr commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+    no longer exist if we treat node deletion as confirming permanent
+    partition. If we do not, the pod GC controller must not force delete
+    pods.
+* It must be possible for an administrator to effectively resolve partitions
```
i wonder if we should bump this up to a condition on namespace or if an event would be sufficient.

derekwaynecarr commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+observe partitions, we propose an additional responsibility to the node controller
+or any future controller that attempts to detect partition. The node controller should
+add an additional condition to pods that have been terminated due to a node failing
+to heartbeat that indicates that the cause of the deletion was node partition.
```
+1 -- this is useful to ensure namespaces stuck terminating are for this reason.

vishh commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+period, but never a longer one.
+
+Deleting a pod with grace period 0 is called **force deletion** and will
+update the pod with a `deletionGracePeriodSeconds` of 0, and then immediately
```
Would it make sense to provide a synchronous delete option? That way, grace period can be separated from actual deletion of the object. In most cases, a user might want to gracefully delete and cleanup a pod right away by specifying grace period as `0`.

smarterclayton replied:
Is the use case that you want to "wait" for the deletion to finish?  Right at the end of 1.5 we made the change to have kubelet wait for deletion, and there is a general issue tracking "wait for action" (which could include delete) to make clients better at scripting.  I would expect those to cover the client side aspects of waiting for deletion within the bounds of our API design philosophy.

vishh replied:
As of now, setting graceperiod to `0` implicitly deletes the pod object from etcd. Instead of having an implicit API, why not make inline deletes explicit? A user can request a graceperiod of `0` and still have the pod object be deleted by the kubelet. If a user wants to delete a pod object right away from etcd, they can then set an explicit parameter to achieve that.

smarterclayton replied:
Would not be backwards compatible, so at best would be something we do in a v2.

vishh commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+A persistent volume that references a strongly consistent storage backend
+like AWS EBS, GCE PD, OpenStack Cinder, or Ceph RBD can rely on the storage
+API to prevent corruption of the data due to simultaneous access by multiple
+clients. However, many commonly deployed storage technologies in the
```
Question: Are these storage technologies not suitable for clustered deployments then?

smarterclayton replied:
They are broadly deployed technologies that benefit from centralized control (because they offer no centralized guarantees themselves).  Most production workloads in the world today run on iSCSI, FibreChannel, or NFS devices in clustered forms - but everyone builds their own solutions to handle clustering (Pacemaker, etc).  Building the affordances in kube to allow them to be safely used is important, even if Kube itself may not contain the functionality to make them safe.

vishh commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+* All existing controllers in the system must be limited to signaling pod
+  termination (starting graceful deletion), and are not allowed to force
+  delete a pod.
+  * The node controller will no longer be allowed to force delete pods -
```
What will happen when the nodes go offline?

kargakis replied:
https://github.com/kubernetes/kubernetes.github.io/pull/1757/files#diff-1fd98dd83a72174d81fbe353fdb248baR62

vishh replied:
@kargakis posted a comment in https://github.com/kubernetes/kubernetes.github.io/pull/1757/files#diff-1fd98dd83a72174d81fbe353fdb248baR62

vishh commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+would be able to leverage a number of systems including but not limited to:
+
+* Cloud control plane APIs such as machine force shutdown
+* Additional agents running on each host to force kill process or trigger reboots
```
why is this necessary? Is kubelet is robust enough, this should not be necessary and we do want kubelet to be robust in general

smarterclayton replied:
Example would be network fencing.  Kubelet may wedge (we cannot prevent wedges), but an independent process might be fine (like kube-proxy).  That independent process could sever the network connection to storage.

I don't think this is required, but it's an example.  There are lots of sophisticated fencing tools today - I'm moderately biased to leaving the door open for their use even if it's a "problem left to someone else to solve".

vishh replied:
As long as it is not part of the default setup, I have no issues.

smarterclayton replied:
I would expect this to be at the level of cloud provider controller or higher, but higher level fencing would probably depend on your setup and not be a "turn on by default" for metal.

vishh commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+* Network routers, backplane switches, software defined networks, or system firewalls
+* Storage server APIs to block client access
+
+to appropriately limit the ability of the partitioned system to impact the cluster.
```
This seems to be a lot of work to get poorly designed software to work in clusters. What about the cost of maintaining such a complex controller? It seems to have a lot of extension points?

vishh replied:
My concern is with third party applications making incorrect assumptions about kube node software. I hope these extensions are managing *other* infrastructure only.

smarterclayton replied:
So I have familiarity with some forms of fencers, but not all.  In general, it's just working up the chain of increasingly large hammers until you hear the breaking glass and can confirm that it's broken.  I.e. rack IPMI telling you that it powered down a node (assuming you trust the IPMI controller) or the TOR switch acknowledging it is now dropping all packets to and from a particular MAC / port / etc.  I would agree that these hammers require a good understanding of their weaknesses before using them.

dashpole commented:
*[docs/proposals/pod-safety.md, line 187 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34160#-KTMg2Ol3xhUtKPe8Ien-r3-187:-KXrOgiGDVnz_ZR6LXNl:b-tyzta4) ([raw file](https://github.com/kubernetes/kubernetes/blob/07964895d4c496290edc62dd773384f1c434bed0/docs/proposals/pod-safety.md#L187)):*
<details><summary><i>Previously, dashpole (David Ashpole) wrote…</i></summary><blockquote>

Ack
</blockquote></details>

The kubelet must not delete a pod until all resources assigned to it are cleaned up.  This includes processes, containers, volumes, and network (and probably more).

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/34160)*
<!-- Sent from Reviewable.io -->

smarterclayton commented:
Moving to https://github.com/kubernetes/community/pull/124, I will close this once I respond and update that PR with the latest comments from vish and ashpole

smarterclayton commented:
Agree that kubelet should cleanup resources of pods before cleaning up the pod from the API.  Will enshrine that in the proposal.

beekhof commented on docs/proposals/pod-safety.md:
```diff
@@ -0,0 +1,431 @@
+
+If the kubelet crashes during the termination process, it will restart the
+termination process from the beginning (grace period is reset). This ensures
+that a process is always given **at least** grace period to terminate cleanly.
```
"at least" may turn out to be problematic.

In other contexts cluster managers usually prefer "at most" as this allows them to continue to make reasonably accurate assumptions even if whole datacenters become inaccessible from one another (in which case the blessed side might reasonably want to assume that the non-blessed side will have killed the relevant pods after some interval).

Deciding which side gets blessed and how is a tangential conversation, for now it's probably only important whether this is a relevant usecase to consider.

smarterclayton replied:
The fencer would be the one who could provide "at most" guarantees.  At the core, we can't get around consensus requirements, so an administrator who wanted to provide "at most" semantics would fence the other data center.  In the current system it would be possible for someone who wants to make an "at most" decision even without strong consensus (observed termination) could still force delete the pods.

smarterclayton replied:
Not sure what you mean by blessed and non-blessed part (terminology mismatch).  The API server is providing serializable guarantees, so there is only progress or not-progress w.r.t. to recorded changes.  The node controller will not do anything if split (can't make changes), but even then it simply requests termination.  The kubelet will observe the request and try to reconcile it, but not it will not handle behavior.

The fencer is allowed to force delete if it can:

1. Reach the master (not partitioned)
2. Guarantee total isolation of the process from the rest of the cluster (storage, network, quantum interference with other processes, whatever)

We definitely will have to consider sets of fencing (storage, network, and then power management), so 2 above is more nuanced than just "force delete" - it might be updating the PV record (as described elsewhere in this doc) as well as force deleting the pod.

If the fencer is wrong, or unsure, the fencer can't force delete the pod.

vishh commented:
@smarterclayton LGTM.

janetkuo commented:
![](https://github.trello.services/images/mini-trello-icon.png) [PetSet](https://trello.com/c/WtVQEMNA/26-petset)

smarterclayton commented:
Marking keep-open until I apply the last comments to the community PR to clarify fencing scope.

smarterclayton commented:
All comments applied to the community doc, please give the second commit a once over and then we'll merge (since this was landed in 1.5).
